src/python/pants/backend/python/goals/run_pex_binary.py

```diff
@@ -85,16 +86,18 @@ async def create_pex_binary_run_request(
         ),
     )
 
-    merged_digest = await Get(
-        Digest,
-        MergeDigests(
-            [
-                pex.digest,
-                local_dists.pex.digest,
-                local_dists.remaining_sources.source_files.snapshot.digest,
-            ]
-        ),
-    )
+    input_digests = [
+        pex.digest,
+        local_dists.pex.digest,
+        # Note regarding inline mode: You might think that the sources don't need to be copied
+        # into the chroot when using inline sources. But they do, because some of them might be
+        # codegenned, and those won't exist in the inline source tree. Rather than incurring the
+        # complexity of figuring out here which sources were codegenned, we copy everything.
+        # The inline source roots precede the chrooted ones in PEX_EXTRA_SYS_PATH, so the inline
+        # sources will take precedence and their copies in the chroot will be ignored.
+        local_dists.remaining_sources.source_files.snapshot.digest,
```

Comment:
I had this concern earlier, but couldn't substantiate it. Now I can.

If we're generating sources as submodules of a parent module, Python will use the parent's path for submodule lookup. So the in-repo path is the only path considered and it fails to import the codegened module :cry:

Reply:
The "fix" is to opt-out of the in-repo running, so everything is in the sandbox. Not great but not fatal.

Reply:
> I can maybe try to play around with namespace packages

Blasting every `__init__.py` up the chain with https://packaging.python.org/en/latest/guides/packaging-namespace-packages/#pkgutil-style-namespace-packages worked.

Tricky Tricky.

Reply:
Yes, the only sane way to solve this is to have the relevant packages be namespace packages. This can be a caveat we document, but since codegen use is not super-common, and needing to run inline is not super-common, I think that's OK.

Reply:
Yeah, I think for my upcoming proposal it'll be documented with an escape hatch to run-in-sandbox (like this). Then you just have to choose between the two.

FWIW my use case might not be sooooo uncommon. I'm compiling a C extension module and I do that in my plugin via codegen. This will be true for generated assets as well (like `HTTPSource` ones). So users will inevitably hit it IMO.

Reply:
> Yes, the only sane way to solve this is to have the relevant packages be namespace packages. This can be a caveat we document, but since codegen use is not super-common, and needing to run inline is not super-common, I think that's OK.

Not quite... the reason this works in the sandbox is because we merge colliding PYTHONPATH entries into a single entry. It would be possible to do something similar in this mode, by using exclusively a sandbox entry if not doing so would result in packages being split across PYTHONPATH entries.

Reply:
That does not qualify as sane IMO. The whole point of this change is to allow you to force something to run in the repo if it needs to. Things do not generally need to. I don't think we need to switch the default here, for example. But if you do need to we shouldn't silently not do what you asked.

Reply:
I'm very of the mind that this PR only fills the gap of not being able to `./pants run random_file.py` in-repo. To me, running the `pex_binary`` actually runs the PEX binary (and there is no "sandbox" vs in-repo toggle).

That being said, this particular issue will travel to the `./pants run random_file.py` implementation.

Reply:
https://github.com/pantsbuild/pants/pull/15849